# Market Segmentation with Clustering - Lab

## Introduction

In this lab, you'll use your knowledge of clustering to perform market segmentation on a real-world dataset!

## Objectives

In this lab you will: 

- Use clustering to create and interpret market segmentation on real-world data 

## Getting Started

In this lab, you're going to work with the [Wholesale customers dataset](https://archive.ics.uci.edu/ml/datasets/wholesale+customers) from the UCI Machine Learning datasets repository. This dataset contains data on wholesale purchasing information from real businesses. These businesses range from small cafes and hotels to grocery stores and other retailers. 

Here's the data dictionary for this dataset:

|      Column      |                                               Description                                              |
|:----------------:|:------------------------------------------------------------------------------------------------------:|
|       FRESH      |                    Annual spending on fresh products, such as fruits and vegetables                    |
|       MILK       |                               Annual spending on milk and dairy products                               |
|      GROCERY     |                                   Annual spending on grocery products                                  |
|      FROZEN      |                                   Annual spending on frozen products                                   |
| DETERGENTS_PAPER |                  Annual spending on detergents, cleaning supplies, and paper products                  |
|   DELICATESSEN   |                           Annual spending on meats and delicatessen products                           |
|      CHANNEL     | Type of customer.  1=Hotel/Restaurant/Cafe, 2=Retailer. (This is what we'll use clustering to predict) |
|      REGION      |            Region of Portugal that the customer is located in. (This column will be dropped)           |



One benefit of working with this dataset for practice with segmentation is that we actually have the ground-truth labels of what market segment each customer actually belongs to. For this reason, we'll borrow some methodology from Supervised Learning and store these labels separately, so that we can use them afterwards to check how well our clustering segmentation actually performed. 

Let's get started by importing everything we'll need.

In the cell below:

* Import `pandas`, `numpy`, and `matplotlib.pyplot`, and set the standard alias for each. 
* Use `numpy` to set a random seed of `0`.
* Set all matplotlib visualizations to appear inline.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

np.random.seed(0)

Now, let's load our data and inspect it. You'll find the data stored in `'wholesale_customers_data.csv'`. 

In the cell below, load the data into a DataFrame and then display the first five rows to ensure everything loaded correctly.

In [3]:
raw_df = pd.read_csv('wholesale_customers_data.csv')

raw_df.head()

Channel  Region  Fresh  Milk  Grocery  Frozen  Detergents_Paper  Delicassen
0        2       3  12669  9656     7561     214              2674        1338
1        2       3   7057  9810     9568    1762              3293        1776
2        2       3   6353  8808     7684    2405              3516        7844
3        1       3  13265  1196     4221    6404               507        1788
4        2       3  22615  5410     7198    3915              1777        5185

Now, let's go ahead and store the `'Channel'` column in a separate variable, and then drop both the `'Channel'` and `'Region'` columns. Then, display the first five rows of the new DataFrame to ensure everything worked correctly. 

In [5]:
channels = raw_df['Channel']
df = raw_df.drop(['Channel', 'Region'], axis = 1)
df.head()

Fresh  Milk  Grocery  Frozen  Detergents_Paper  Delicassen
0  12669  9656     7561     214              2674        1338
1   7057  9810     9568    1762              3293        1776
2   6353  8808     7684    2405              3516        7844
3  13265  1196     4221    6404               507        1788
4  22615  5410     7198    3915              1777        5185

Now, let's get right down to it and begin our clustering analysis. 

In the cell below:

* Import `KMeans` from `sklearn.cluster`, and then create an instance of it. Set the number of clusters to `2`
* Fit it to the data (`df`) 
* Get the predictions from the clustering algorithm and store them in `cluster_preds` 

In [7]:
from sklearn.cluster import KMeans

In [8]:
k_means = KMeans(n_clusters = 2)
k_means.fit(df)

cluster_preds = k_means.predict(df)

Now, use some of the metrics to check the performance. You'll use `calinski_harabasz_score()` and `adjusted_rand_score()`, which can both be found inside `sklearn.metrics.cluster`. 

In the cell below, import these scoring functions. 

In [9]:
from sklearn.metrics import calinski_harabasz_score, adjusted_rand_score

Now, start with CH score, to get the variance ratio. 

In [10]:
calinski_harabasz_score(df, cluster_preds)

171.68461633384186

Although you don't have any other numbers to compare this to, this is a pretty low score, suggesting that the clusters aren't great. 

Since you actually have ground-truth labels in this case, you use `adjusted_rand_score()` to check how well the clustering performed. Adjusted Rand score is meant to compare two clusterings, which the score can interpret our labels as. This will tell us how similar the predicted clusters are to the actual channels. 

Adjusted Rand score is bounded between -1 and 1. A score close to 1 shows that the clusters are almost identical. A score close to 0 means that predictions are essentially random, while a score close to -1 means that the predictions are pathologically bad, since they are worse than random chance. 

In the cell below, call `adjusted_rand_score()` and pass in `channels` and `cluster_preds` to see how well your first iteration of clustering performed. 

In [11]:
adjusted_rand_score(channels, cluster_preds)

-0.03060891241109425

According to these results, the clusterings were essentially no better than random chance. Let's see if you can improve this. 

### Scaling our dataset

Recall that K-means clustering is heavily affected by scaling. Since the clustering algorithm is distance-based, this makes sense. Let's use `StandardScaler` to scale our dataset and then try our clustering again and see if the results are different. 

In the cells below:

* Import and instantiate [StandardScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html) and use it to transform the dataset  
* Instantiate and fit K-means to this scaled data, and then use it to predict clusters 
* Calculate the adjusted Rand score for these new predictions 

In [12]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaled_df = scaler.fit_transform(df)

In [13]:
scaled_k_means = KMeans(n_clusters = 2)
scaled_k_means.fit(scaled_df)

scaled_preds = scaled_k_means.predict(scaled_df)

In [14]:
adjusted_rand_score(channels, scaled_preds)

0.212843835451224

That's a big improvement! Although it's not perfect, we can see that scaling our data had a significant effect on the quality of our clusters. 

## Incorporating PCA

Since clustering algorithms are distance-based, this means that dimensionality has a definite effect on their performance. The greater the dimensionality of the dataset, the greater the total area that we have to worry about our clusters existing in. Let's try using Principal Component Analysis to transform our data and see if this affects the performance of our clustering algorithm. 

Since you've already seen PCA in a previous section, we will let you figure this out by yourself. 

In the cells below:

* Import [PCA](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html) from the appropriate module in sklearn 
* Create a `PCA` instance and use it to tranform our scaled data  
* Investigate the explained variance ratio for each Principal Component. Consider dropping certain components to reduce dimensionality if you feel it is worth the loss of information 
* Create a new `KMeans` object, fit it to our pca-transformed data, and check the adjusted Rand score of the predictions it makes. 

**_NOTE:_** Your overall goal here is to get the highest possible adjusted Rand score. Don't be afraid to change parameters and rerun things to see how it changes. 

In [15]:
from sklearn.decomposition import PCA

In [16]:
pca = PCA()

In [17]:
scaled_pca_df = pca.fit_transform(scaled_df)

In [19]:
pca.explained_variance_ratio_

array([0.44082893, 0.283764  , 0.12334413, 0.09395504, 0.04761272,
       0.01049519])

In [20]:
kmeans_pca = KMeans(n_clusters = 2)
kmeans_pca.fit(scaled_pca_df)
kmeans_pca_preds = kmeans_pca.predict(scaled_pca_df)

adjusted_rand_score(channels, kmeans_pca_preds)

0.19214043169327583

In [27]:
for comps in range(1, 6, 1):
    pca_iter = PCA(n_components=comps, random_state=7)
    scaled_iter = pca_iter.fit_transform(scaled_df)
    kmeans_pca_iter = KMeans(n_clusters=2, random_state=7)
    kmeans_pca_iter.fit(scaled_iter)
    kmeans_pca_iter_preds = kmeans_pca_iter.predict(scaled_iter)
    print(f'The adjusted random score for the K-means with {comps} principal components \n retained is: {adjusted_rand_score(channels, kmeans_pca_iter_preds)}')

The adjusted random score for the K-means with 1 principal components 
 retained is: 0.20696616374562318
The adjusted random score for the K-means with 2 principal components 
 retained is: 0.23084287036169227
The adjusted random score for the K-means with 3 principal components 
 retained is: 0.23084287036169227
The adjusted random score for the K-means with 4 principal components 
 retained is: 0.19957248499831937
The adjusted random score for the K-means with 5 principal components 
 retained is: 0.212843835451224


**_Question_**:  What was the Highest Adjusted Rand Score you achieved? Interpret this score, and determine the overall quality of the clustering. Did PCA affect the performance overall?  How many principal components resulted in the best overall clustering performance? Why do you think this is?

Write your answer below this line:
_______________________________________________________________________________________________________________________________

The highest Adjusted Random Score I achieved was 0.231 with standard scaled and PCA transformed data with number of components = 2. The clustering is better than random, but could still be improved. The first two components together explain 72% of the variance, which apparently is enough to sufficiently model the clusters.

## Optional (Level up) 

### Hierarchical Agglomerative Clustering

Now that we've tried doing market segmentation with K-means clustering, let's end this lab by trying with HAC!

In the cells below, use [Agglomerative clustering](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html) to make cluster predictions on the datasets we've created, and see how HAC's performance compares to K-mean's performance. 

**_NOTE_**: Don't just try HAC on the PCA-transformed dataset -- also compare algorithm performance on the scaled and unscaled datasets, as well! 

In [28]:
from sklearn.cluster import AgglomerativeClustering

In [ ]:
#try on unscaled data, scaled data, and PCA scaled data
#try different linkage types - Ward, complete, average, single

In [30]:
hac = AgglomerativeClustering()
hac.fit(df)
hac_pred = hac.labels_
adjusted_rand_score(channels, hac_pred)

-0.01923156414375716

In [31]:
hac_scaled = AgglomerativeClustering()
hac_scaled.fit(scaled_df)
hac_pred = hac_scaled.labels_
adjusted_rand_score(channels, hac_pred)

0.022565317001188977

In [32]:
hac_scaled_pca = AgglomerativeClustering()
hac_scaled_pca.fit(scaled_pca_df)
hac_pred = hac_scaled_pca.labels_
adjusted_rand_score(channels, hac_pred)

0.022565317001188977

In [34]:
for comps in range(1, 6, 1):
    pca_iter = PCA(n_components=comps, random_state=7)
    scaled_iter = pca_iter.fit_transform(scaled_df)
    hac_pca_iter = AgglomerativeClustering()
    hac_pca_iter.fit(scaled_iter)
    hac_pca_iter_preds = hac_pca_iter.labels_
    print(f'The adjusted random score for the HAC with {comps} principal components \n retained is: {adjusted_rand_score(channels, hac_pca_iter_preds)}')

The adjusted random score for the HAC with 1 principal components 
 retained is: 0.4864773957379437
The adjusted random score for the HAC with 2 principal components 
 retained is: 0.0459127111971714
The adjusted random score for the HAC with 3 principal components 
 retained is: 0.02763975237440555
The adjusted random score for the HAC with 4 principal components 
 retained is: 0.04822381910875346
The adjusted random score for the HAC with 5 principal components 
 retained is: 0.24781414655368145


In [37]:
pca_1 = PCA(n_components=1, random_state=7)
pc_scaled = pca_1.fit_transform(scaled_df)
for linkage in ["ward", "complete", "average", "single"]:
    hac_iter = AgglomerativeClustering(linkage = linkage)
    hac_iter.fit(pc_scaled)
    hac_iter_preds = hac_iter.labels_
    print(f'The adjusted random score for the HAC with {linkage} linkage is: \n{adjusted_rand_score(channels, hac_iter_preds)}')
    

The adjusted random score for the HAC with ward linkage is: 
0.4864773957379437
The adjusted random score for the HAC with complete linkage is: 
0.053441339729341335
The adjusted random score for the HAC with average linkage is: 
0.053441339729341335
The adjusted random score for the HAC with single linkage is: 
0.004987197295533421


## Summary

In this lab, you used our knowledge of clustering to perform a market segmentation on a real-world dataset. You started with a cluster analysis with poor performance, and then implemented some changes to iteratively improve the performance of the clustering analysis!